<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
    <script src="shared/jquery.js" type="text/javascript"></script>
    <script src="shared/shiny.js" type="text/javascript"></script>
    <link rel="stylesheet" type="text/css" href="shared/shiny.css"/>
    <link rel="stylesheet" type="text/css" href="shared/slider/css/jquery.slider.min.css"/>
    <script src="shared/slider/js/jquery.slider.min.js"></script>
    <link rel="stylesheet" type="text/css" href="shared/bootstrap/css/bootstrap.min.css"/>
    <script src="shared/bootstrap/js/bootstrap.min.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <script src="http://d3js.org/d3.v3.min.js" charset="utf-8"></script>
    <script src="plot.js" type="text/javascript"></script>
    <link rel="stylesheet" type="text/css" href="plot.css"/>
    <link rel="stylesheet" type="text/css" href="shared/bootstrap/css/bootstrap-responsive.min.css"/>
    <title>Shiny GAM</title>
  </head>
  <body>
      <div id="networkParameters" class="js-output">
          <script>
              network = {
                  available: false,
                  hasReactionsAsEdges: false,
                  hasReactionsAsNodes: false,
                  usesRpairs: false
              };
          </script>
      </div>
      <div id="moduleParameters" class="js-output">
          <script>
              module = {
                  available: false
              };
          </script>
      </div>
    <div class="container">
      <div class="row">
        <div class="span12" style="padding: 10px 0px;">
          <h1>Shiny GAM</h1>
        </div>
      </div>
      <div class="row">
          <div class="span12">
              <div class="tabbable">
                  <ul class="nav nav-tabs" id="tab-panel-main">
                      <li class="active"> <a href="#tab-main" data-toggle="tab">Work</a> </li>
                      <li> <a href="#tab-help" data-toggle="tab">Help</a> </li>
                      <li> <a href="#tab-about" data-toggle="tab">About</a> </li>
                  </ul>
                  <div class="tab-content">
                      <div class="tab-pane active" id="tab-main">
                          <div class="row">
                              <div class="span3">
                                  <form class="well">
                                      <label class="control-label" for="network">
                                          Select an organism: 
                                          <a id="select-organism-tooltip" class="help-tooltip" href="#" 
                                              title="You have to select an organism">
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                      <select id="network">
                                          <option value="Mouse musculus" selected="selected">Mouse musculus</option>
                                          <option value="Homo sapiens">Homo sapiens</option>
                                      </select>
                                      <label>Select a TSV-file with gene p-values</label>
                                      <input id="geneDE" type="file" accept="text/tab-separated-values"/>
                                      <div id="geneDEProgress" class="progress progress-striped active shiny-file-input-progress">
                                          <div class="bar"></div>
                                          <label></label>
                                      </div>
                                      <label>Select a TSV-file with metabolite p-values</label>
                                      <input id="metDE" type="file" accept="text/tab-separated-values"/>
                                      <div id="metDEProgress" class="progress progress-striped active shiny-file-input-progress">
                                          <div class="bar"></div>
                                          <label></label>
                                      </div>
                                      <label class="control-label" for="reactionsAs">How to interpret reactions:</label>
                                      <select id="reactionsAs">
                                          <option value="edges" selected="selected">As edges</option>
                                          <option value="nodes">As nodes</option>
                                      </select>
                                      <div data-display-if="input.reactionsAs == 'nodes'">
                                          <label class="checkbox" for="collapseReactions">
                                              <input id="collapseReactions" type="checkbox" checked="checked"/>
                                              <span>Collapse reactions</span>
                                          </label>
                                      </div>
                                      <div data-display-if="input.reactionsAs != 'nodes'">
                                          <label class="checkbox" for="useRpairs">
                                              <input id="useRpairs" type="checkbox" checked="checked"/>
                                              <span>Use RPAIRs</span>
                                          </label>
                                      </div>
                                      <br/>
                                      <button id="preprocess" type="button" class="btn action-button">Make network</button>
                                  </form>
                              </div>
                              <div class="span9">
                                  <div class="row">
                                      <div class="span4">
                                          <h3>Genomic data</h3>
                                          <div id="geneDESummary" class="shiny-html-output"></div>
                                          <div id="geneDETable" class="shiny-html-output"></div>
                                      </div>
                                      <div class="span4">
                                          <h3>Metabolic data</h3>
                                          <div id="metDESummary" class="shiny-html-output"></div>
                                          <div id="metDETable" class="shiny-html-output"></div>
                                      </div>
                                  </div>
                                  <h3>Network summary</h3>
                                  <div id="networkSummary" class="shiny-html-output"></div>
                                  <div data-display-if="network.available">
                                      <a id="downloadNetwork" class="btn shiny-download-link" href="" target="_blank">Download XGMML</a><br>
                                  </div>
                              </div>
                          </div>
                          <div id="module-panel" class="row">
                              <div class="span3">
                                  <form class="well">
                                      <label for="geneFDR">FDR for genes</label>
                                      <input id="geneFDR" type="number" value="1e-06" min="1e-100" max="1" step="1e-200"/>
                                      <label for="metFDR">FDR for metabolites</label>
                                      <input id="metFDR" type="number" value="1e-06" min="1e-100" max="1" step="1e-200"/>
                                      <label for="absentMetScore">Score for absent metabolites</label>
                                      <input id="absentMetScore" type="number" value="-20"/>
                                      <label for="absentRxnScore">Score for absent reactions</label>
                                      <input id="absentRxnScore" type="number" value="-10"/>
                                      <br/>
                                      <button id="find" type="button" class="btn action-button">Find module</button>
                                      <br/>
                                      <div data-display-if="network.hasReactionsAsEdges && network.usesRpairs">
                                          <label class="checkbox" for="addTransPairs">
                                              <input id="addTransPairs" type="checkbox"/>
                                              <span>Add trans- edges</span>
                                          </label>
                                      </div>
                                      <div data-display-if="network.hasReactionsAsNodes">
                                          <label class="checkbox" for="addMetabolitesForReactions">
                                              <input id="addMetabolitesForReactions" type="checkbox"/>
                                              <span>Add all metabolites for reactions</span>
                                          </label>
                                          <label class="checkbox" for="addInterconnections">
                                              <input id="addInterconnections" type="checkbox"/>
                                              <span>Add interconnections</span>
                                          </label>
                                          <label class="checkbox" for="removeHangingNodes">
                                              <input id="removeHangingNodes" type="checkbox"/>
                                              <span>Remove hanging nodes</span>
                                          </label>
                                          <label class="checkbox" for="removeSimpleReactions">
                                              <input id="removeSimpleReactions" type="checkbox"/>
                                              <span>Remove simple reactions</span>
                                          </label>
                                      </div>
                                      <br/>
                                      <br/>
                                  </form>
                                  <h3>Module summary</h3>
                                  <div id="moduleSummary" class="shiny-html-output"></div>
                                  <div data-display-if="module.available">
                                      <a id="downloadModule" class="btn shiny-download-link" href="" target="_blank">Download XGMML</a>
                                  </div>
                              </div>
                              <div class="span9">
                                  <div id="module" class="graph-output"></div>
                              </div>
                          </div>
                      </div>
                      <div class="tab-pane" id="tab-help">
                          <div class="row">
                              <div class="span9 offset3">
                                  <p>
                                  This web-application provides a way to analyse transcriptional and/or 
                                  metabolic data in a context of biochemical reactions.
                                  </p>

                                  <p>
                                  First, select an organism, so that only reactions possible
                                  in the organism are used.
                                  </p>

                                  <p>
                                  Then upload available data: gene and/or
                                  metabolite differential expression data in a
                                  tab-separated format. The first line of these
                                  files should be a header with column names.
                                  Columns should include "ID", "pval" and
                                  "logFC": identificator of a gene (RefSeq mRNA
                                  transcript ID or Entrez ID), P-value for the
                                  gene and a logarithm of a fold-change
                                  correspondingly.  You can download example
                                  data for <a href="/publications/supp_materials/GAM/gene.de.M1.M2.tsv">genes</a>
                                  and <a href="/publications/supp_materials/GAM/met.de.M1.M2.tsv">metabolites</a>.
                                  As you upload your files summary of them will
                                  appear, so you can check that it was parsed
                                  correctly.
                                  </p>

                                  <p>
                                  Select an interpretation of reactions in a
                                  network: either as edges or as nodes.  For
                                  interpretation of reactions as edges
                                  network is made in a
                                  following way. For each pair of metabolites
                                  with multiple reaction connections between
                                  them only one with minimal p-value is kept.
                                  Then all the remaining reactions and
                                  metabolites that take part in these reactions
                                  are combined into a network as edges and nodes
                                  correspondingly. A possible option is to
                                  consider not all cross-connections between
                                  substrates and products for a reaction, but
                                  only those that make up a KEGG reaction pair
                                  (RPAIR, see <a href=http://www.genome.jp/kegg/reaction/>KEGG REACTION</a> for details).
                                  </p>

                                  <p>
                                  If reaction are interpreted as nodes then
                                  both metabolites and reactions are added as
                                  nodes to the network. Edges
                                  are added between a metabolite and a reaction
                                  if the metabolite takes part in this reaction.
                                  Then, as an option, groups of reactions
                                  having at least one common metabolite and the
                                  same most significant gene can be collapsed
                                  into single nodes.
                                  </p>

                                  <p>
                                  Generally, all this options lead to similar 
                                  results. We recommend to use default option
                                  values as this make a network simpler and analysis
                                  faster.
                                  </p>

                                  <p>
                                  After making a network you can find a
                                  connected module that contains the most
                                  significantly changed genes and reactions.
                                  Internally this is done by firstly scoring nodes
                                  and edges based on their P-values such that
                                  positive scores correspond to significant
                                  P-values and negative scores correspond to
                                  insignificant changes. Then the problem
                                  of finding a connected subgraph with maximum
                                  summary weight is solved. As this problem
                                  is NP-hard and solving it may take a lot of 
                                  time there is a one minute time limit for 
                                  running solver.
                                  By changing FDR values and scores for absent
                                  metabolites and reactions you can control
                                  size of the module. We recommend to start
                                  from default values and then gradually change
                                  them depending on the results. 
                                  </p>

                                  <p>
                                  There are couple of post-processing steps that 
                                  are available. If in the network reactions are
                                  edges and RPAIRs are used then trans- 
                                  connections can be added. When reactions are
                                  nodes following operations could be applied:
                                  <ul>
                                      <li> Adding all metabolites that are part
                                      of any rection in the module. </li>

                                      <li> Adding all reactions that aren't present
                                      in the module but are directly connected
                                      to two metabolites from the module.</li>

                                      <li> Removing all metabolites with no data
                                      and only one connection in the module </li>

                                      <li> Replacing reaction nodes with edges 
                                      for reactions that connect only two
                                      metabolites in the module and these two
                                      metabolites are on different sides of 
                                      the reaction. </li>
                                  </ul>
                                  </p>

                                  <p>
                                  Anytime you can download the module in a XGMML
                                  format and, for example, explore it in Cytoscape.
                                  You can download GAM's VizMap style for Cytoscape
                                  <a href="/publications/supp_materials/GAM/GAM_VizMap.xml">here</a>.
                                  </p>
                              </div>
                          </div>
                      </div>
                      <div class="tab-pane" id="tab-about">
                          <div class="row">
                              <div class="span9 offset3">
                                  <p> By Alexey Sergushichev.  </p>
                                  <p> GAM version: </p>
                                  <p> Links <p>
                                  <p> Powered by Shiny. </p>
                              </div>
                          </div>
                      </div>
                  </div>
              </div>
          </div>
      </div>
    </div>
    <div id="showModulePanel" class="js-output"></div>
  </body>
</html>
