<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
    <script src="shared/jquery.js" type="text/javascript"></script>
    <script src="shared/shiny.js" type="text/javascript"></script>
    <link rel="stylesheet" type="text/css" href="shared/shiny.css"/>
    <link rel="stylesheet" type="text/css" href="shared/slider/css/jquery.slider.min.css"/>
    <script src="shared/slider/js/jquery.slider.min.js"></script>
    <link rel="stylesheet" type="text/css" href="shared/bootstrap/css/bootstrap.min.css"/>
    <script src="shared/bootstrap/js/bootstrap.min.js"></script>
    <link rel="stylesheet" type="text/css" href="shared/bootstrap/css/bootstrap.min.css"/>

    <script src="http://caseyjhol.github.io/bootstrap-select/javascripts/bootstrap-select.js"></script>
    <link rel="stylesheet" type="text/css" href="http://caseyjhol.github.io/bootstrap-select/stylesheets/bootstrap-select.css"/>


    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <script src="http://d3js.org/d3.v3.min.js" charset="utf-8"></script>
    <script src="plot.js" type="text/javascript"></script>
    <link rel="stylesheet" type="text/css" href="plot.css"/>
    <link rel="stylesheet" type="text/css" href="shared/bootstrap/css/bootstrap-responsive.min.css"/>
    <title>Shiny GAM</title>
  </head>
  <body>
      <div id="networkParameters" class="js-output">
          <script>
              network = {
                  available: false,
                  hasReactionsAsEdges: false,
                  hasReactionsAsNodes: false,
                  hasGenes: false,
                  usesRpairs: false
              };
          </script>
      </div>
      <div id="moduleParameters" class="js-output">
          <script>
              module = {
                  available: false
              };
          </script>
      </div>
    <div class="container">
      <div class="row">
        <div class="span12" style="padding: 10px 0px;">
          <h1>Shiny GAM</h1>
        </div>
      </div>
      <div class="row">
          <div class="span12">
              <div class="tabbable">
                  <ul class="nav nav-tabs" id="tab-panel-main">
                      <li class="active"> <a href="#tab-main" data-toggle="tab">Work</a> </li>
                      <li> <a href="#tab-help" data-toggle="tab">Help</a> </li>
                      <li> <a href="#tab-about" data-toggle="tab">About</a> </li>
                  </ul>
                  <div class="tab-content">
                      <div class="tab-pane active" id="tab-main">
                          <div class="row">
                              <div class="span3">
                                  <form class="well">
                                      <p class="form-header">Construct a network</p>
                                      <label class="control-label" for="network">
                                          Select an organism: 
                                          <a id="select-organism-tooltip" class="help-tooltip" href="#" 
                                              title="Select an organism that you want to analyse. If there is no one that you want contact the author (see About tab).">
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                      <select id="network" class="selectpicker">
                                          <option value="Mouse musculus" selected="selected">Mouse musculus</option>
                                          <option value="Homo sapiens">Homo sapiens</option>
                                      </select>
                                      <label>
                                          File with gene DE data: 
                                          <a id="upload-gene-data-tooltip" class="help-tooltip" href="#" 
                                              title='File with gene differential expression data shoud be tab-separated with header and contain columns "ID" and "pval".'>
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                      <input id="geneDE" type="file" accept="text/tab-separated-values"/>
                                      <label>
                                          File with metabolic DE data: 
                                          <a id="upload-met-data-tooltip" class="help-tooltip" href="#" 
                                              title='File with metabolic differential expression data shoud be tab-separated with header and contain columns "ID" and "pval".'>
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                      <input id="metDE" type="file" accept="text/tab-separated-values"/>
                                      <label class="control-label" for="reactionsAs">
                                          How to interpret reactions:
                                          <a id="reactionAs-tooltip" class="help-tooltip" href="#" 
                                              title='Reactions can be either edges connecting substrates with products or nodes connected to them. Reactions as edges makes analysis more flux-centric.'>
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                      <select id="reactionsAs">
                                          <option value="edges" selected="selected">As edges</option>
                                          <option value="nodes">As nodes</option>
                                      </select>
                                      <div data-display-if="input.reactionsAs == 'nodes'">
                                          <label class="checkbox" for="collapseReactions">
                                              <input id="collapseReactions" type="checkbox" checked="checked"/>
                                              <span>Collapse reactions
                                                  <a id="collapseReactions-tooltip" class="help-tooltip" href="#" 
                                                      title='Collapse reaction nodes with the same enzyme.'>
                                                      <i class="icon-question-sign"></i>
                                                  </a>
                                              </span>
                                          </label>
                                      </div>
                                      <div data-display-if="input.reactionsAs != 'nodes'">
                                          <label class="checkbox" for="useRpairs">
                                              <input id="useRpairs" type="checkbox" checked="checked"/>
                                              <span>Use RPAIRs
                                                  <a id="useRpairs-tooltip" class="help-tooltip" href="#" 
                                                      title='Connect only those substrate-product pairs that make up a KEGG RPAIR.'>
                                                      <i class="icon-question-sign"></i>
                                                  </a>
                                              </span>
                                          </label>
                                      </div>
                                      <button id="preprocess" type="button" class="btn action-button">Make network</button>
                                  </form>
                              </div>
                              <div class="span9">
                                  <div class="row">
                                      <div class="span4">
                                          <h3>Genomic data</h3>
                                          <div id="geneDESummary" class="shiny-html-output"></div>
                                          <div id="geneDETable" class="shiny-html-output"></div>
                                      </div>
                                      <div class="span4">
                                          <h3>Metabolic data</h3>
                                          <div id="metDESummary" class="shiny-html-output"></div>
                                          <div id="metDETable" class="shiny-html-output"></div>
                                      </div>
                                  </div>
                                  <div class="bottom-buffer">
                                      <h3>Network summary</h3>
                                      <div id="networkSummary" class="shiny-html-output"></div>
                                      <div data-display-if="network.available">
                                          <a id="downloadNetwork" class="btn shiny-download-link" href="" target="_blank">Download XGMML</a><br>
                                      </div>
                                  </div>
                              </div>
                          </div>
                          <div id="module-panel" class="row top-buffer">
                              <div class="span3">
                                  <form class="well">
                                      <p class="form-header">Discover modules</p>
                                      <div data-display-if="network.hasGenes">
                                          <label for="geneFDR">FDR for reactions
                                              <a id="geneFDR-tooltip" class="help-tooltip" href="#" 
                                                  title='Increasing/decreasing this value makes adding reactions to a module harder/easier.'>
                                                  <i class="icon-question-sign"></i>
                                              </a>
                                          </label>
                                          <input id="geneFDR" type="number" value="1e-06" min="1e-100" max="1" step="1e-200"/>
                                      </div>

                                      <label for="metFDR">FDR for metabolites
                                          <a id="metFDR-tooltip" class="help-tooltip" href="#" 
                                              title='Increasing/decreasing this value makes adding metabolites to a module harder/easier.'>
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                      <input id="metFDR" type="number" value="1e-06" min="1e-100" max="1" step="1e-200"/>

                                      <label for="absentMetScore">Score for absent metabolites
                                          <a id="absentMetScore-tooltip" class="help-tooltip" href="#" 
                                              title='Increasing/decreasing this value makes adding metabolites with no data to a module harder/easier.'>
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                      <input id="absentMetScore" type="number" value="-20"/>

                                      <div data-display-if="!network.hasGenes">
                                          <label for="absentRxnScore">Score for absent reactions
                                              <a id="absentRxnScore-tooltip" class="help-tooltip" href="#" 
                                                  title='Increasing/decreasing this value makes adding reactions with no data to a module harder/easier (usable only when there is no gene expression data).'>
                                                  <i class="icon-question-sign"></i>
                                              </a>
                                          </label>
                                          <input id="absentRxnScore" type="number" value="-10"/>
                                      </div>

                                      <label class="control-label" for="solver">Select a solver:
                                          <a id="selectSolver-tooltip" class="help-tooltip" href="#" 
                                              title='Multiple MWCS solvers can be used to find modules. Check the help page for details.'>
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>

                                      <select id="solver">
                                          <option value="heinz" selected="selected">Heinz</option>
                                          <option value="mwcs">MWCS (under development)</option>
                                          <option value="fastHeinz">Heuristic search</option>
                                      </select>

                                      <div data-display-if="$('#solver').val() == 'mwcs'">
                                          <label for="mwcsTimeLimit">MWCS solver time limit (seconds):
                                          <a id="selectSolver-tooltip" class="help-tooltip" href="#" 
                                              title="It's an approximate maximum time MWCS solver is allowed to run. After hitting the time limit it will print the best solurion found so far.">
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                          <input id="mwcsTimeLimit" type="number" value="10" min="0" max="120"/>
                                      </div>

                                      <div data-display-if="$('#solver').val() == 'heinz'">
                                          <label for="heinzTimeLimit">Heinz time limit (seconds):
                                          <a id="selectSolver-tooltip" class="help-tooltip" href="#" 
                                              title="It's an approximate maximum time heinz solver is allowed to run. After hitting the time limit it will just stop searching outputing no module.">
                                              <i class="icon-question-sign"></i>
                                          </a>
                                      </label>
                                          <input id="heinzTimeLimit" type="number" value="60" min="0" max="240"/>
                                      </div>
                                      <button id="find" type="button" class="btn action-button">Find module</button>
                                      <div data-display-if="network.hasReactionsAsEdges && network.usesRpairs">
                                          <label class="checkbox" for="addTransPairs">
                                              <input id="addTransPairs" type="checkbox"/>
                                              <span>Add trans- edges
                                                  <a id="addTransPairs-tooltip" class="help-tooltip" href="#" 
                                                      title='Add trans- RPAIR edges.'>
                                                      <i class="icon-question-sign"></i>
                                                  </a>
                                              </span>
                                          </label>
                                      </div>
                                      <div data-display-if="network.hasReactionsAsNodes">
                                          <label class="checkbox" for="addMetabolitesForReactions">
                                              <input id="addMetabolitesForReactions" type="checkbox"/>
                                              <span>Add all metabolites for reactions
                                                  <a id="addMetabolitesForReactions-tooltip" class="help-tooltip" href="#" 
                                                      title='Add all metabolites that are part of any reaction in the module.'>
                                                      <i class="icon-question-sign"></i>
                                                  </a>
                                              </span>
                                          </label>
                                          <label class="checkbox" for="addInterconnections">
                                              <input id="addInterconnections" type="checkbox"/>
                                              <span>Add interconnections
                                                  <a id="addInterconnections-tooltip" class="help-tooltip" href="#" 
                                                      title="Addg all reactions that aren't present in the module but are directly connected to two metabolites from the module.">
                                                      <i class="icon-question-sign"></i>
                                                  </a>
                                              </span>
                                          </label>
                                          <label class="checkbox" for="removeHangingNodes">
                                              <input id="removeHangingNodes" type="checkbox"/>
                                              <span>Remove hanging nodes
                                                  <a id="removeHangingNodes-tooltip" class="help-tooltip" href="#" 
                                                      title='Remove all metabolites with no data and only one connection in the module.'>
                                                      <i class="icon-question-sign"></i>
                                                  </a>
                                              </span>
                                          </label>
                                          <label class="checkbox" for="removeSimpleReactions">
                                              <input id="removeSimpleReactions" type="checkbox"/>
                                              <span>Remove simple reactions
                                                  <a id="removeSimpleReactions-tooltip" class="help-tooltip" href="#" 
                                                      title='Replacing reaction nodes with edges for reactions that connect only two metabolites in the module and these two metabolites are on different sides of the reaction.'>
                                                      <i class="icon-question-sign"></i>
                                                  </a>
                                              </span>
                                          </label>
                                      </div>
                                  </form>
                                  <h3>Module summary</h3>
                                  <div id="moduleSummary" class="shiny-html-output"></div>
                                  <div data-display-if="module.available">
                                      <a id="downloadModule" class="btn shiny-download-link" href="" target="_blank">Download XGMML</a>
                                  </div>
                              </div>
                              <div class="span9">
                                  <div id="module" class="graph-output"></div>
                              </div>
                          </div>
                      </div>
                      <div class="tab-pane" id="tab-help">
                          <div class="row">
                              <div class="span9 offset3">
                                  <h3>Quick start</h3>

                                  <p>
                                  Reload the page, so that all parameters will be set to default.
                                  </p>

                                  <p>
                                  Download example
                                  data for <a href="/publications/supp_materials/GAM/gene.de.M1.M2.tsv">genes</a>
                                  and <a href="/publications/supp_materials/GAM/met.de.M1.M2.tsv">metabolites</a>.
                                  </p>

                                  <p>
                                  Upload downloaded files as TSV-files with gene and metabolite p-values respectively.
                                  You'll see a summary of the data:
                                  </p>

                                  <div class="row">
                                      <div class="span8">
                                          <div class="example-panel">
                                              <div class="row">
                                                  <div class="span4">
                                                      <h3>Genomic data</h3>
                                                      <div class="shiny-html-output shiny-bound-output"><ul>
                                                              <li> length :  14276 </li>
                                                              <li> ID type :  RefSeq </li>
                                                          </ul>
                                                      </div>
                                                      <div class="shiny-html-output shiny-bound-output"><!-- html table generated in R 3.0.2 by xtable 1.7-1 package -->
                                                          <table class="data table table-bordered table-condensed">
                                                              <tbody><tr> <th>  </th> <th> ID </th> <th> pval </th> <th> logFC </th>  </tr>
                                                                  <tr> <td align="right"> 1 </td> <td> NM_021274 </td> <td> 1.489938e-50 </td> <td> -12.292868 </td> </tr>
                                                                  <tr> <td align="right"> 2 </td> <td> NM_013653 </td> <td> 1.098282e-46 </td> <td> -10.498761 </td> </tr>
                                                                  <tr> <td align="right"> 3 </td> <td> NM_010501 </td> <td> 3.872073e-46 </td> <td> -12.144405 </td> </tr>
                                                                  <tr> <td align="right"> 4 </td> <td> NM_001177351 </td> <td> 1.126691e-40 </td> <td>  -9.145330 </td> </tr>
                                                                  <tr> <td align="right"> 5 </td> <td> NM_008332 </td> <td> 4.930544e-40 </td> <td>  -9.710362 </td> </tr>
                                                                  <tr> <td align="right"> 6 </td> <td> NM_145137 </td> <td> 9.538526e-37 </td> <td>  12.364258 </td> </tr>
                                                      </tbody></table></div>
                                                  </div>
                                                  <div class="span4">
                                                      <h3>Metabolic data</h3>
                                                      <div class="shiny-html-output shiny-bound-output"><ul>
                                                              <li> length :  2119 </li>
                                                              <li> ID type :  HMDB </li>
                                                          </ul>
                                                      </div>
                                                      <div class="shiny-html-output shiny-bound-output"><!-- html table generated in R 3.0.2 by xtable 1.7-1 package -->
                                                          <table class="data table table-bordered table-condensed">
                                                              <tbody><tr> <th>  </th> <th> ID </th> <th> pval </th> <th> logFC </th>  </tr>
                                                                  <tr> <td align="right"> 1 </td> <td> HMDB00634 </td> <td> 5.077866e-34 </td> <td> -3.17008 </td> </tr>
                                                                  <tr> <td align="right"> 2 </td> <td> HMDB00620 </td> <td> 5.077866e-34 </td> <td> -3.17008 </td> </tr>
                                                                  <tr> <td align="right"> 3 </td> <td> HMDB02092 </td> <td> 5.077866e-34 </td> <td> -3.17008 </td> </tr>
                                                                  <tr> <td align="right"> 4 </td> <td> HMDB00749 </td> <td> 5.077866e-34 </td> <td> -3.17008 </td> </tr>
                                                                  <tr> <td align="right"> 5 </td> <td> HMDB10720 </td> <td> 2.273393e-32 </td> <td> -2.77394 </td> </tr>
                                                                  <tr> <td align="right"> 6 </td> <td> HMDB03407 </td> <td> 2.273393e-32 </td> <td> -2.77394 </td> </tr>
                                                      </tbody></table></div>
                                                  </div>
                                              </div>
                                          </div>
                                      </div>
                                  </div>
                                  <br>

                                  <p>
                                  Click "Make network" button and wait a few seconds till the network is constructed.
                                  When it happens you'll be automatically scrolled down to a next panel.
                                  You can scroll back up and check network summary or download it in XGMML:
                                  </p>
                                  <div class="row">
                                      <div class="span3">
                                          <div class="example-panel bottom-buffer">
                                              <h3>Network summary</h3>
                                              <div class="shiny-html-output shiny-bound-output"><ul>
                                                      <li> number of nodes :  1541 </li>
                                                      <li> number of edges :  1488 </li>
                                                  </ul>
                                              </div>
                                              <span class="btn shiny-download-link shiny-bound-output">Download XGMML</span><br>
                                          </div>
                                      </div>
                                  </div>
                                  <br>

                                  <p>
                                  Click "Find module" button to find a module in the network. You can preview it on the right panel:
                                  </p>
                                  <div><img src="module.png"/></div>

                                  <br>

                                  <p>
                                  Click "Download XGMML" button in the "Module summary section" to download the module in XGMML format for further analysis in Cytoscape:
                                  </p>
                                  <div class="row">
                                      <div class="span3">
                                          <div class="example-panel bottom-buffer">
                                              <h3>Module summary</h3>
                                              <div class="shiny-html-output shiny-bound-output"><ul>
                                                      <li> number of nodes :  34 </li>
                                                      <li> number of edges :  36 </li>
                                                  </ul>
                                              </div>
                                              <span class="btn shiny-download-link shiny-bound-output">Download XGMML</span>
                                          </div>
                                      </div>
                                  </div>

                                  <h3>Details</h3>
                                  <p>
                                  This application is a web-interface to R-package 
                                  <a href="bioconductor/GAM" target="_blank">GAM</a> and
                                  provides a way to analyse transcriptional and/or 
                                  metabolic differential expression data 
                                  in a context of biochemical reactions.
                                  </p>

                                  <p>
                                  Analysis consists of creating a network 
                                  describing connections between
                                  metabolites and reactions specific
                                  to the data and finding a connected module
                                  that contains most significant changes.
                                  </p>

                                  <h4>Constructing a network</h4>

                                  <p>
                                  To construct a network you first have to select 
                                  an organism, so that only reactions possible
                                  in the organism are used.
                                  </p>

                                  <p>
                                  Available data (gene and/or
                                  metabolite DE data) should 
                                  be in a tab-separated format. The first line of these
                                  files should be a header with column names.
                                  Columns should include "ID", "pval" and
                                  "logFC": identifier (RefSeq mRNA
                                  transcript ID or Entrez ID for genes and HMDB 
                                  or KEGG ID for metabolites), p-value and 
                                  a logarithm of a fold-change.
                                  As you upload your files you can see summary
                                  of them appear, so you can check that it was parsed
                                  correctly. Even if you have only one type
                                  of data it's still possible to do an analysis.
                                  </p>

                                  <p>
                                  After uploading gene DE data it's converted
                                  to reaction DE data by selecting a gene
                                  coding enzyme for a reaction with minimal p-value.
                                  If there is a gene DE data then all reactions
                                  without p-values or discarded as having no
                                  expressed enzymes. We recommend to exclude
                                  genes with low expression prior uploading.
                                  </p>

                                  <p>
                                  Then you should select an interpretation of 
                                  reactions in the network: either as 
                                  edges or as nodes.  For
                                  interpretation of reactions as edges
                                  the network is constructed in a
                                  following way. For each pair of metabolites
                                  with multiple reaction connections between
                                  them only one with a minimal p-value is kept.
                                  Then all the remaining reactions and
                                  metabolites that take part in these reactions
                                  are combined into a network as edges and nodes
                                  respectively. A possible option is to
                                  consider not all cross-connections between
                                  substrates and products for a reaction, but
                                  only those that make up a KEGG reaction pair
                                  (RPAIR, see <a href=http://www.genome.jp/kegg/reaction/>KEGG REACTION</a> for details).
                                  </p>

                                  <p>
                                  If reaction are interpreted as nodes then
                                  both metabolites and reactions are added as
                                  nodes to the network. Edges
                                  are added between a metabolite and a reaction
                                  if the metabolite takes part in this reaction.
                                  Then, as an option, groups of reactions
                                  having at least one common metabolite and the
                                  same most significant gene can be collapsed
                                  into single nodes.
                                  </p>

                                  <p>
                                  Generally, all this options lead to similar 
                                  results. We recommend to use default option
                                  values as this make a network simpler and analysis
                                  faster.
                                  </p>

                                  <h4>Finding a module</h4>

                                  <p>
                                  After making a network you can find a
                                  connected module that contains the most
                                  significantly changed genes and reactions.
                                  Internally this is done by firstly scoring nodes
                                  and edges based on their p-values such that
                                  positive scores correspond to significant
                                  p-values and negative scores correspond to
                                  insignificant changes. Then the problem
                                  of finding a connected subgraph with maximum
                                  summary weight (maximum-weight connected subgraph —
                                  MWCS — problem) is solved. 
                                  </p>

                                  <p>
                                  This site supports three solvers: 
                                  <ul>
                                      <li><a href='http://www.mi.fu-berlin.de/w/LiSA/Heinz'>Heinz solver</a> 
                                      is a default solver, as it works for both 
                                      interpretation of reactions.</li>
                                      <li>Using MWCS solver can provide more interactive
                                      experience because when reaching time limit it
                                      outputs the best solution found. </li>
                                      <li>Heuristic search is solver from R-package 
                                      BioNet.</li> 
                                  </ul>
                                  As the last two solvers don't support edge scores
                                  they are not available when reactions are
                                  interpreted as edges.
                                  </p>

                                  <p>
                                  By changing FDR values and scores for absent
                                  metabolites and reactions you can control
                                  size of the module. We recommend to start
                                  from default values and then gradually change
                                  them depending on the results. 
                                  </p>

                                  <p>
                                  There are couple of post-processing steps that 
                                  are available. If in the network reactions are
                                  edges and RPAIRs are used then trans- 
                                  connections can be added. When reactions are
                                  nodes following operations could be applied:
                                  <ul>
                                      <li> Adding all metabolites that are part
                                      of any reaction in the module. </li>

                                      <li> Adding all reactions that aren't present
                                      in the module but are directly connected
                                      to two metabolites from the module.</li>

                                      <li> Removing all metabolites with no data
                                      and only one connection in the module. </li>

                                      <li> Replacing reaction nodes with edges 
                                      for reactions that connect only two
                                      metabolites in the module and these two
                                      metabolites are on different sides of 
                                      the reaction. </li>
                                  </ul>
                                  </p>

                                  <p>
                                  You can download the module in a XGMML
                                  format and, for example, explore it in Cytoscape.
                                  You can download GAM's VizMap style for Cytoscape
                                  <a href="/publications/supp_materials/GAM/GAM_VizMap.xml">here</a>.
                                  </p>
                              </div>
                          </div>
                      </div>
                      <div class="tab-pane" id="tab-about">
                          <div class="row">
                              <div class="span9 offset3">
                                  <p> By Alexey Sergushichev (<a href="mailto:asergushichev@wustl.edu">asergushichev@wustl.edu</a>)</p>
                                  <p> This site must not be used for commercial purposes.<p>
                                  <p> You can find source code of this site <a href="https://bitbucket.org/assaron/gam">here</a>.</p>
                                  <div id="GAMVersion" class="shiny-html-output"></div>
                                  <p> Links <p>
                                  <p> Powered by Shiny. </p>
                              </div>
                          </div>
                      </div>
                  </div>
              </div>
          </div>
      </div>
    </div>
    <div id="showModulePanel" class="js-output"></div>
  </body>
</html>
