<!--
%\VignetteIndexEntry{Genomic and metabolomic integrated analysis tutorial}
%\VignetteEngine{knitr::knitr}
-->
```{r eval=TRUE, echo=FALSE, message=FALSE}
require(knitr)
opts_chunk$set(error=TRUE, cache=TRUE, dev="svg")
```

# Genomic and metabolomic integrated analysis tutorial

This is a tutorial that describes how 
to find the most
significant changed module between two states in the cellular reaction network
having metabolic mass spectrometry
and gene expression data.

## Quick start

This section shows the basics of the GAM package.

First, load packages and data. Package `igraph` contains methods to print summary for networks.
Package `mouseMacrophages` contains data from experiment that we will use as an example.

```{r message=FALSE}
library("GAM")
library("GAM.db")
library("igraph")
data("examplesGAM")
```


Load differential expression data for mouse macrophages:
```{r}
library(RCurl)
gene.de.M0.M1 <- read.csv(text=getURL("http://artyomovlab.wustl.edu/publications/supp_materials/GAM/Ctrl.vs.MandLPSandIFNg.gene.de.tsv"), sep="\t")
met.de.M0.M1 <- read.csv(text=getURL("http://artyomovlab.wustl.edu/publications/supp_materials/GAM/Ctrl.vs.MandLPSandIFNg.met.de.tsv"), sep="\t")
```

Load network for mouse. The package contains prebuilt networks for mouse and
human (`kegg.human.network`). For details, how networks can be constructed see
section [Constructing network](#constructing-network) for details.

```{r}
library("GAM.networks")
data("kegg.mouse.network")
```


Next step is to make an experiment set that contains all the data 
necessary to find modules. You can do it by calling a function
`makeExperimentSet` which accepts as arguments network,
available differential expression data, and type of IDs that is
used in this data. One part of preprocessing is to fit p-values distribution
to beta-uniform model, by default `makeEXperimentSet` makes histogram and QQ-plot,
so you can check if it fits good enough.
```{r cache=TRUE, error=FALSE, message=FALSE, fig.width=4, fig.height=4}
es.re <- makeExperimentSet(network=kegg.mouse.network,
                           met.de=met.de.M0.M1,
                           gene.de=gene.de.M0.M1,
                           reactions.as.edges=T)
```

Result contains gene and metabolite differential expression with
converted IDs, p-values for reactions and subnetwork with only
those reactions that have expressed genes for enzymes.
```{r}
head(es.re$gene.de)
head(es.re$met.de)
head(es.re$rxn.de)
es.re$subnet
```

To be able to find modules `GAM` need a solver for maximum-wight
connected subgraph problem. [Heinz 2](https://software.cwi.nl/cwisoftware/software/heinz) 
is used for this purpose. See section [MWCS solvers](#mwcs-solvers) for additional information.
```{r}
solver <- heinz2.solver("/usr/local/lib/heinz2/heinz", timeLimit=30)
print(solver)
```

Set FDRs and score for absent metabolites. Using these parameters size of the module
can be controlled: larger FDRs make larger modules. Score for absnet metabolites controls
how many such metabolites will be in the module.
```{r}
met.fdr <- 3e-5
rxn.fdr <- 3e-5
absent.met.score <- -20
```

And now we can finally find most significant module in our network (if heinz is installed).
```{r chache=TRUE, error=FALSE, eval=FALSE}
module.re <- findModule(es.re,
                        met.fdr=met.fdr,
                        rxn.fdr=rxn.fdr,
                        absent.met.score=absent.met.score,
                        solver=solver)
```
```{r chache=TRUE, error=FALSE}
module.re
```

You can plot this module using command `plotNetwork`.
```{r fig.width=8,fig.height=8,error=FALSE,message=FALSE}
plotNetwork(module.re)
```

There are a few functions available for post-processing, see section [Postprocessing](#postprocessing).


After that, you can save resulting module into pdf, XGMML
or few other formats.
```{r error=FALSE}
saveModule(module.re,
           paste0("module.M0.M1.re", 
                  ".mf=", met.fdr,
                  ".rf=", rxn.fdr,
                  ".ms=", absent.met.score),
           types=c("pdf", "XGMML")
)
```

There is a Cytoscape VizMap style supplied with the package that can be used 
for visualization:
```{r echo=FALSE}
system.file("GAM_VizMap.xml", package="GAM")
```
     
## Interpreting reactions as nodes

Alternative approach is to represent both metabolites and reactions as nodes.
In this case, metabolite and reaction are connected if the metabolite takes part
in the reaction.

```{r chache=TRUE, error=FALSE, message=FALSE}
es.rn <- makeExperimentSet(network=kegg.mouse.network,
                           met.de=met.de.M0.M1,
                           gene.de=gene.de.M0.M1,
                           reactions.as.edges=F,
                           plot=F)
es.rn$subnet
```

FDRs should be set a little stricter:
```{r}
met.fdr=c(5e-6)
rxn.fdr=c(5e-6)
```

And finding modules:
```{r eval=FALSE}
module.rn <- findModule(es.rn,
                        met.fdr=met.fdr,
                        rxn.fdr=rxn.fdr,               
                        solver=solver)
```

```{r}
module.rn
```

Module:
```{r fig.width=8, fig.height=8, error=FALSE,message=FALSE}
plotNetwork(module.rn)
```

Again, for available post-processing functions see section [Postprocessing](#postprocessing)

## Constructing network <a name="constructing-network"></a>

Network construction uses KEGG database. `GAM.db` package goes with
neccessary parts of KEGG in `kegg.db` object:

```{r}
library("GAM.db")
data(kegg.db)
names(kegg.db)
```

It includes mapping from reaction to its enzymes and mapping from enzym to its coding genes
(now just for mouse and human):
```{r}
head(kegg.db$rxn2enz)
head(kegg.db$enz2gene)
```

Connections between metabolites on different sides of reactions:
```{r}
head(kegg.db$net)
```

Names for reactions and metabolites. In analysis names of reaction genes are
usually used as reaction labels, not reaction names.
```{r}
head(kegg.db$rxn2name)
head(kegg.db$met2name)
```

Metabolites and reactions to mask. Masked metabolites are non-organic molecules, 
some generic molecules and some others. Masked reactions are reactions that 
are parts of multi-step reactions.
```{r}
head(kegg.db$mets2mask)
head(kegg.db$rxns2mask)
```

Some of metabolites, e.g. anomeric ones, can be collaps into one node.
```{r}
head(kegg.db$mets2collapse)
```

You can make a network for a specific organism from all this data with
`makeKeggNetwork` function:
```{r cache=TRUE}
kegg.mouse.network <- makeKeggNetwork(kegg.db, "MMU")
```

Mouse network is actually a list of triples `met.x`, `rxn` and `met.y`,
each of its elements says that compounds `met.x`
and `met.y` are on different sides of the reaction `rxn`.
```{r}
head(kegg.mouse.network$graph.raw)
```

All of these data can be modified for specific needs. For example,
one can add reaction converting cis-Aconitate to Itaconate which
was not present in KEGG as for September 2013:
```{r cache=TRUE}
kegg.db$net <- rbind(kegg.db$net, c("C00417", "RXIRG1", "C00490", "RPXIRG1", "main"))
kegg.db$rxn2enz <- rbind(kegg.db$rxn2enz, c("RXIRG1", "Irg1"))
kegg.db$enz2gene <- rbind(kegg.db$enz2gene, c("Irg1", "16365", "MMU"))
kegg.mouse.network <- makeKeggNetwork(kegg.db, "MMU")
```


## Postprocessing <a name="postprocessing"></a>

Postprocessing options depend on what interpretation of reactions was used.

#### Reactions as edges

You can add edges for trans- pairs of metabolites in reactions:
```{r}
module.re <- addTransEdges(module.re, es.re)
```

#### Reactions as nodes

You can add all metabolites taking part in every reaction in the module:
```{r}
module.rn <- addMetabolitesForReactions(module.rn, es.rn)
```

Add all reactions that connect any metabolites in the module:
```{r}
module.rn <- addInterconnections(module.rn, es.rn)
```

Add attribute for log-fold change normalized across nodes of the
same type. Thist attribute is used in `plotNetwork` function and 
Cytoscape visual style.
```{r}
module.rn <- addNormLogFC(module.rn)
```

Remove hanging nodes without p-values:
```{r}
module.rn <- removeHangingNodes(module.rn)
```

Replace reaction nodes for simple one-to-one reactions with
edges:
```{r}
module.rn <- simplifyReactionNodes(module.rn, es.rn)
```

Copy reaction attributes for edges adjacent to 
reaction nodes:
```{r}
module.rn <- expandReactionNodeAttributesToEdges(module.rn)
```

## MWCS solvers <a name="mwcs-solvers"></a>

For now `GAM` supports three solvers: `heinz.solver`, `heinz2.solver` and `gmwcs.solver`.

**Heinz solver** is a solver first described in Dittrich et al. 2008. It allows
to solve the problem of finding maximum-weighted connected subgraph even if
there are positive edge weights. However, it  searches only for tree solutions.
Heinz is able to find the exact optimal solution very fast in practice. The
last version for x64 linux machine can be downloaded 
[here](http://homepages.cwi.nl/~klau/data/heinz_1.68.tgz).

**Heinz 2 solver** is described in El-Kebir&Klau 2014. It solves MWCS problem
where only nodes are scored. It cannot be used when reactions are mapped to
edges and there is differential expression for genes available. For instances
appearing in GAM it works more robust than Heinz in finding a good solution
within a limited amount of time. The last version
x64 linux machine can be downloaded 
[here](https://software.cwi.nl/software/heinz).

The gmwcs solver is an experimental solver for the problem of finding
maximum-weighted connected subgraph with weighted edges, not limited to tree
solutions (in opposite to heinz). It is available at
https://github.com/alexloboda/gmwcs-solver. Following Haouari et al, 2013 it
reduces MWCS problem to integer linear programming problem which is
consequently solved by IBM ILOG CPLEX. It also shows more robust performance
than heinz when the time is limited.

If you have probles making this solvers work, you can try to use a `GAM` web-interface at
http://artyomovlab.wustl.edu/shiny/gam.

## References

1. Beisser, D.; Klau, G. W.; Dandekar, T.; Mueller, T. and Dittrich, M. (2009). BioNet an R-package for the Functional Analysis of Biological Networks. Bioinformatics 26, 1129-1130.
2. Dittrich, Marcus T., et al. "Identifying functional modules in protein-protein interaction networks: an integrated exact approach." Bioinformatics 24.13 (2008): i223-i231.
3. Beisser, Daniela, et al. "Integrated pathway modules using time-course metabolic profiles and EST data from Milnesium tardigradum." BMC systems biology 6.1 (2012): 72.
4. Csardi, Gabor, and Tamas Nepusz. "The igraph software package for complex network research." InterJournal, Complex Systems 1695.5 (2006).
5. Shannon, Paul, et al. "Cytoscape: a software environment for integrated models of biomolecular interaction networks." Genome research 13.11 (2003): 2498-2504.
6. El-Kebir, M., & Klau, G. W. Solving the Maximum-Weight Connected Sub-graph Problem to Optimality. arXiv preprint arXiv:1409.5308 (2014). 

